Prevent web apps/app preview usage if a domain has too many mobile UCRs

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
https://dimagi-dev.atlassian.net/browse/SAAS-14934

This work was prompted by out of memory errors observed on mobile webworkers. These out of memory errors were caused by restore requests coming from a domain with lots of mobile UCRs (UCRs referenced in applications given the MOBILE_UCR feature flag is enabled).  App aware restores would help, but formplayer restores are not app-aware by design, so the only solution to reducing the risk of these problematic restores is to restrict web apps usage if a mobile UCR count is too high.

The user experience certainly isn't pretty, but this is not a SaaS owned feature flag so the priority is to prevent the possibility of bringing down mobile webworkers. I've attached the two screenshots where this is applicable, though it is possible that I also need to add this check/redirection to the FormplayerPreviewSingleApp view. I started out by just handling the two main entry points into formplayer that I know of. Here are the screenshots to show what the user will see when attempting to access web apps or live preview.

![Screenshot from 2023-10-30 20-45-29](https://github.com/dimagi/commcare-hq/assets/15785053/294155d5-3d0c-444f-8ec5-4f9bfbf2313f)

![Screenshot from 2023-10-30 20-45-11](https://github.com/dimagi/commcare-hq/assets/15785053/be5a598f-a4f7-4121-97a2-d9848d8630c7)

The initial value is set to 300 based on my first pass on the typical number of mobile UCRs by domains. It is likely that we will want to lower this number in the near future though.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
MOBILE_UCR
## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested locally, have thorough automated tests for the utility responsible for determining if a domain should be blocked, and will run through QA.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Automated tests have been written to ensure the utility behaves as expected.
### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-5808

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
prevent user from using web apps/app preview if too many mobile UCRs

snopoke reviewed (commented):
Seems OK as a temporary measure - would be great to surface this issue earlier in the app building process.

biyeun commented on corehq/apps/cloudcare/templates/preview_app/too_many_ucrs_preview.html:
```diff
@@ -0,0 +1,10 @@
+{% load i18n %}
+
+{% block body %}
+<div>
```
not necessary to wrap a `<p>` in an unstyled `<div>`

I would recommend styling the `div` with the css class `alert alert-warning`

gherceg replied:
![Screenshot from 2023-10-31 10-34-35](https://github.com/dimagi/commcare-hq/assets/15785053/13624d87-0214-478e-b4c7-04b4f7ad03fe)
Updated!

gherceg replied:
a1f41126477e116b471718cc09c8149a07f107d6

biyeun commented on corehq/apps/cloudcare/templates/too_many_ucrs.html:
```diff
@@ -0,0 +1,13 @@
+
+{% block page_name %}{% trans "You are currently blocked from using Web Apps" %}{% endblock %}
+{% block page_content %}
+<div>
```
I would remove the unstyled `div` here and add `class="error"` to the `<p>` tag

gherceg replied:
a1f41126477e116b471718cc09c8149a07f107d6

biyeun reviewed (commented):
backend functionality looks good, just some comments to address on the front end

gherceg commented:
> Seems OK as a temporary measure - would be great to surface this issue earlier in the app building process.

Agreed. Another longer term resolution is to move everyone onto V2 of mobile UCRs since that isn't as memory intensive as V1.

gherceg committed:
update styling

esoergel commented:
Have you been able to determine whether rolling this out will have an immediate effect on any existing projects?

This system is very abrupt and harsh - I'm guessing what will happen when this gets triggered is that there will be a mad dash to understand the error and address it as quickly as possible.  I don't know that there's an easy way to make it less abrupt (cloning an app could easily trigger it), but in my opinion it'd be prudent to plan for a quick out that doesn't require a deploy.

Say a USH project triggers this error - we'll probably negotiate and say "okay we'll fix it, give us a week to reduce UCR usage without breaking workflows, but turn it back on now!"  At present, that'd require a code deploy, but alternatives would be making the limit dynamically configurable per project (would this fit with the existing rate limiter code?), or making a whitelist feature flag to temporarily exempt a domain from the limit.

esoergel commented:
Just had another idea - it's only built apps that count towards this limit, right?  One option would be to raise this as a validation error.  That means you couldn't make any new builds until you were under the threshold, but it shouldn't affect app users.

gherceg commented:
Thanks for the feedback @esoergel. In hindsight I absolutely should have created a spec for this first and shared it around before starting any development here so that is my bad.

As for your questions, with the 300 UCR threshold, only one project is impacted, with another project pretty close to that threshold (which to your point, could be set over the threshold if they cloned an app). However the important thing to convey here is that once your mobile UCR count reaches a certain number (unclear what number this is), web apps and live preview are virtually unusable because any attempt at using it will trigger a restore which will timeout/fail.

Ideally we would find the magic number of UCRs that doesn't cause a failed restore, but the idea is to start here. Knowing that, does that change your perspective on this work at all?

esoergel commented:
Good to hear that this only impacts one project.  I don't have any issues with imposing a limit on this, even one much smaller than 300 if that's what's needed to ensure platform stability.  My quibble is more with the enforcement mechanism than the limit.  For web apps based projects, turning off web apps means the project stops - it's the same level of severity as HQ going down.  I think we should give fair warning before resorting to that.  Restores timing out is also obviously very bad for the same reasons, so it's good to see this work to address a common cause of that.

Another consideration that occurred to me is that it might be hard for projects to remove mobile UCRs with live preview and webapps disabled, as they won't be able to test anything until they are under the limit, though I don't have a good suggestion for that issue.

orangejenny reviewed (dismissed):
> For web apps based projects, turning off web apps means the project stops - it's the same level of severity as HQ going down.

Re-emphasizing this. A USH project hitting this limit would be an automatic P1, and it would be very hard for us to explain why we took the entire system down instead of alerting the internal team to immediately address the issue.

This conversation is going in a good direction, and I see from the ticket that this particular limit was selected after due diligence, but I think the enforcement mechanism here is an unacceptable risk.

I also don't think that local testing and no automated tests is sufficient here. When the risk is taking down Web Apps, there should at least be a second set of eyes testing this.

gherceg commented:
> Just had another idea - it's only built apps that count towards this limit, right? One option would be to raise this as a validation error. That means you couldn't make any new builds until you were under the threshold, but it shouldn't affect app users.

The UCR count is based on the number of primary/current applications in a domain, so excluding builds. I'm basing this on the query used to obtain apps to pull UCRs from in fixture provider code (https://github.com/dimagi/commcare-hq/blob/gh/cases/remove-es-docs/corehq/apps/app_manager/dbaccessors.py#L256-L261).

I do like the spirit of this suggestion though. Similar to Simon's point of surfacing this earlier in the build process, we would ideally prevent the user from adding a mobile UCR reference to an app once they have hit the limit, which leaves web apps/live preview functional, gives the user a clear idea of what needs to be done to resolve the issue, and most importantly prevents anyone from panicking.

I wasn't thinking about how to prevent projects from getting into this state in the first place, but was instead just attempting to stop the immediate issue. However I think a better approach for resolving the immediate issue is to communicate with the most problematic projects directly, and work with them to reduce their mobile UCR count since it is beneficial for both us and them (again, web apps/live preview are in unusable states). We can then start work on safer long term solutions to prevent this problem, including potentially limiting UCR references and completing the migration of the mobile UCR restore version to V2.

Thanks again for the feedback. I had a gut feeling this was not a good approach so your comments are validating. Feel free to add any additional comments here, but I'm going to close this one out.

gherceg commented:
Another alternative to this PR is to just put up a banner or some text informing the user that their mobile UCR count is high, and they are prone to experiencing disruptions when using web apps/live preview. This could also point them to documentation on how to resolve this.

esoergel commented:
@gherceg thanks, those all sound like reasonable next steps, in my opinion.  It definitely sounds useful to impose limits here to prevent frequent timeouts, as long as we can get the mechanism right.

> The UCR count is based on the number of primary/current applications in a domain, so excluding builds. I'm basing this on the query used to obtain apps to pull UCRs from in fixture provider code.

Good find - I think that has considerable room for improvement, if perhaps out of scope for what you're doing here.  It does indeed look like this pulls fixtures for every "current" app in the domain, as you say, which is wrong in a few ways.  First off, for web apps restores, I'd expect to only consider applications which:
* Have "web apps" enabled
* Have a released build
* Are accessible to the user

And I'd expect it to use the reports defined in the latest released build, _not_ the current version.  I'm surprised that hasn't caused problems, in fact (maybe I'm misreading something?) - if a report is removed in the current version of the app, then it'd go missing from restores, even if the live version still uses it.

Live preview conflicts with that, because that needs the latest saved version of the application.  I can't think of a way to reconcile that without separate user sandboxes for live preview and for web apps (or ideally, actually app-aware restores).

Still, there might be room for some optimization here, like syncing only the first set of reports for users who don't have access to live preview, so only app builders get the mammoth set of reports currently sent down.

orangejenny commented:
@esoergel Idle thought: it's also surprising to me that app preview doesn't use an app-aware restore, since it's app-specific. Any sense if changing that would significantly improve app preview performance?

esoergel commented:
@orangejenny whether it improves performance would depend a lot on the app and the project.  If the project uses mobile UCR and has more than one app on the domain, then certainly.  Otherwise I doubt they'd see much of an improvement, as case load is the same regardless of app.

I assumed app preview and webapps share a user sandbox - is that not the case?  I think these are the possibilities:
* One sandbox per user (current situation, I think)
* Up to two sandboxes per user, one for web apps and one for live preview (this would solve the versioning issue, and let us focus web apps restores on released builds accessible to the user)
* One sandbox for web apps, and one sandbox per app for live preview (lower mobile UCR restore load for app builders)
* One sandbox per app for both web apps and live preview (true app aware restores)

The big disadvantage of more sandboxes per user is that we'd be storing case data in duplicate in each sandbox, which is wasted storage, and might have issues with liveliness (changes in one app not immediately impacting another).  However, those issues only impact users that use more than one app on a domain.

I don't have a great sense of how feasible any of this would be though - changing it would be primarily on the formplayer side, I expect, perhaps also with some changes to the load balancer.

gherceg committed:
create ALLOW_WEB_APPS_RESTRICTION feature flag

gherceg committed:
add check for ALLOW_WEB_APPS_RESTRICTION feature flag

- also renames util method to check if domain should be restricted
- adds tests for the util method should_restrict_web_apps_usage

gherceg committed:
store MAX_MOBILE_UCR_LIMIT in settings instead of constant

this adds another method of altering behavior on live environments
without requiring a deploy

gherceg committed:
rename too_many_ucrs to more general block_web_apps

gherceg commented:
I wrote up a [brief spec](https://docs.google.com/document/d/16O6tLMZIyBSfNvRB5QXdy5Qf7nww3zXn0XAzPTO7kk0/edit) to figure out improvements that could be made here that would satisfy the concerns mentioned in feedback. I've also started a section to track what other improvements can and should be made to this area of code.

> if a report is removed in the current version of the app, then it'd go missing from restores, even if the live version still uses it.

Good point. That sounds correct to me, and I've added this into the spec as a followup item.

I'm going to reopen this pull request and push the following improvements:
- create a new feature flag that _needs_ to be enabled on a domain in order for blocking web apps to even be possible. Meaning by default, this will not apply to domains.
- add automated tests to the util responsible for determining if a domain should be limited. ("have to do a lot of db setup for what is a pretty straightforward method" was a pretty terrible excuse in hindsight)

Thanks again for all of the feedback so far.

gherceg committed:
address linting errors

- my new editor wasn't adding a newline at eof, so fixed that
- flake8 was complaining about an unrelated line in cloudcare/urls
being too long, so formatted it differently to resolve

esoergel reviewed (approved):
Thanks for the thoroughness!

gherceg committed:
Move default setting to before localsettings import

gherceg committed:
Merge branch 'master' into gh/web-apps/limit-usage

gherceg commented:
QA passed, but I wanted to flag [this behavior](https://dimagi-dev.atlassian.net/browse/QA-5830?focusedCommentId=299264) here just for transparency. A live preview refresh will not trigger a check to see if the user can access web apps, so in theory a user could add or remove references to UCRs via the app builder, refresh the live preview window, and still have access/not have access when the opposite is actually true. This isn't great UX, but my perspective is this is intended for extreme scenarios, mainly where a domain has already well exceeded the limit, and that followup work will ensure users are made aware of this limit earlier in the app building process to the point where they would rarely, if ever, encounter the actual restriction on web apps.

gherceg committed:
Merge branch 'master' into gh/web-apps/limit-usage

esoergel commented on corehq/apps/cloudcare/utils.py:
```diff
@@ -31,3 +35,26 @@ def webapps_module_case_form(domain, app_id, module_id, case_id, form_id):
     return _webapps_url(domain, app_id, selections=[module_id])
+
+
+def should_restrict_web_apps_usage(domain):
```
It just occurred to me that this would be a good target for caching.  It'll only fire if `ALLOW_WEB_APPS_RESTRICTION` is disabled, so it mostly won't matter, but when that _is_ enabled, iterating through all apps report configs will take a decent amount of time.
